astropy/utils/data.py

```diff
@@ -244,16 +246,49 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
             delete_fds.append(tmp)
             fileobj_new = bz2.BZ2File(tmp.name, mode='rb')
             fileobj_new.read(1)  # need to check that the file is really bzip2
-        except IOError:  # invalid bzip2 file
+        except (IOError) as e:  # invalid bzip2 file
             fileobj.seek(0)
             fileobj_new.close()
+            # raise
         else:
             fileobj_new.seek(0)
             close_fds.append(fileobj_new)
             fileobj = fileobj_new
+    elif signature[:3] == b'\xfd7z':  # xz
+        try:
+            # for Python < 3.3 try backports.lzma; pyliblzma installs as lzma,
+            # but does not support TextIOWrapper
+            if sys.version_info >= (3,3,0):
+                import lzma
+                fileobj_new = lzma.LZMAFile(fileobj, mode='rb')
+            else:
+                from backports import lzma
+                from backports.lzma import LZMAFile
+                # when called with file object, returns a non-seekable instance
+                # need a filename here, too, so have to write the data to a 
+                # temporary file
+                tmp = NamedTemporaryFile("wb", delete=False)
```

Comment:
Use of a `with` clause here will ensure the tmp gets closed even if an except is raised in the interim.